### PR TITLE
Improve code

### PR DESCRIPTION
**Pull request by zhexuany**
Still working on it. Next step is adding more test case to verify the correctness and illuminate corn case. Kill such case via test cases.

---
This change is [Reviewable](https://reviewable.io/reviews/innerr/quotachan/1)

### CODE REVIEW

**Comment by innerr** [comment]
No google c++style, too ugly.

File: chan.h [hunk 1]
```diff
@@ -12,43 +12,59 @@ using std::queue;
...
+  // buffered queue properties
   queue<T> que;
   mutable mutex mtx;
-  condition_variable cond;
+  condition_variable r_cond;
+  condition_variable w_cond;
 
-  const size_t quota;
+  int r_waiting;
+  int w_waiting;
```
Comment (birdstorm) [inline chan.h]: as far as I know, variables in structs and classes in C++ won't initialize to zero by default. 
you should initialize them in constructor beforehand.

Comment (zhexuany) [reply chan.h]: I did.

File: chan.h [hunk 2]
```diff
@@ -11,45 +11,71 @@ using std::mutex;
...
+  inline void Push(const T &v) {
+    unique_lock<mutex> lock(mtx);
+    // blocks until sth is removed since queue is full now.
+    while (que.size() == capacity) {
+      w_waiting++;
+      w_cond.wait(lock);
+      w_waiting--;
     }
 
-    inline bool Pop(T &v) {
```
Comment (innerr) [inline chan.h]: inline friend bool operator << (T &v, Chan &chan)

File: chan.h [hunk 3]
```diff
@@ -11,45 +11,71 @@ using std::mutex;
...
-        {
-            unique_lock<mutex> lock{mtx};
-            que.push(v);
-        }
-        cond.notify_one();
+  Chan(size_t quota_)
+      : capacity(quota_), passed(0), r_waiting(0), w_waiting(0) {}
+
+  // semantic send in golang's channel
+  inline void Push(const T &v) {
```
Comment (innerr) [inline chan.h]: inline friend Chan & operator << (Chan &chan, T &v)

File: chan_test.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  for (int i = 0; i < quota; ++i)
+    jobs.Push(i);
+
+  Chan<int> loadeds(quota);
+
+  vector<thread> loaders(concurrent);
+  for (auto it = loaders.begin(); it != loaders.end(); ++it) {
+    *it = thread([&] {
+      int job = -1;
+      while (jobs.Pop(job)) {
```
Comment (innerr) [inline chan_test.h]: while (job << jobs) {

File: chan_test.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        loadeds.Push(job);
```
Comment (innerr) [inline chan_test.h]: loadeds << job;

**Comment by innerr** [comment]
Added some fun game, check my comments.

File: chan.h (hunk 3 above)
Comment (innerr) [reply chan.h]: capacity does not mean quota

File: chan.h [hunk 6]
```diff
@@ -11,45 +11,71 @@ using std::mutex;
...
 
-    const size_t quota;
-    size_t passed;
+  int r_waiting;
+  int w_waiting;
+  const size_t capacity;
+  size_t passed;
 
 public:
-    Chan(size_t quota_) : quota(quota_), passed(0) {}
```
Comment (innerr) [inline chan.h:24]: What we need:
Chan(size_t quota_, size_t capacity_) : quota(quota_), capacity(capacity_)
When passed == quota, Pop() always return false, means this channel was closed.
When queue.size() == capacity, Push() should wait, till some one call Pop().

File: chan.h [hunk 7]
```diff
@@ -11,45 +11,71 @@ using std::mutex;
...
+template <typename T> class Chan {
+  // buffered queue properties
+  queue<T> que;
+  mutable mutex mtx;
+  condition_variable r_cond;
+  condition_variable w_cond;
 
-    const size_t quota;
-    size_t passed;
+  int r_waiting;
```
Comment (innerr) [inline chan.h]: We dont need this waiting counter, just call notify_all, it works. check it here http://en.cppreference.com/w/cpp/thread/condition_variable/notify_all

**Comment by innerr** [comment]
Still need this constructor:
Chan(size_t quota_, size_t capacity_) : quota(quota_), capacity(capacity_)

**Pull request closed without merging**